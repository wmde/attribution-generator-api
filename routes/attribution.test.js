const setup = require('./__helpers__/setup');
const licenseFactory = require('../__helpers__/licenseFactory');

describe('attribution routes', () => {
  const services = {
    fileData: { getFileData: jest.fn() },
    licenses: { getLicense: jest.fn() },
    licenseStore: { getLicenseById: jest.fn() },
  };
  const fileInfoMock = {
    title: 'File:Apple_Lisa2-IMG_1517.jpg',
    normalizedTitle: 'File:Apple Lisa2-IMG 1517.jpg',
    rawUrl: 'https://upload.wikimedia.org/wikipedia/commons/b/bc/Apple_Lisa2-IMG_1517.jpg',
    wikiUrl: 'https://commons.wikimedia.org/',
    artistHtml:
      '<a href="//commons.wikimedia.org/wiki/User:Rama" title="User:Rama">Rama</a> &amp; Mus√©e Bolo',
    attributionHtml:
      'Photograph by <a href="//commons.wikimedia.org/wiki/User:Rama" title="User:Rama">Rama</a>, Wikimedia Commons, Cc-by-sa-2.0-fr',
  };
  const license = licenseFactory({
    id: 'cc-by-sa-2.5',
    name: 'CC BY-SA 2.5',
    groups: ['cc', 'cc2.5', 'ccby'],
    url: 'https://creativecommons.org/licenses/by-sa/2.5/legalcode',
  });

  let context;

  beforeEach(async () => {
    context = await setup({ services });
  });

  afterEach(async () => {
    await context.destroy();
  });

  describe('GET /attribution/... (unmodified)', () => {
    const file = 'File:Foobar.jpg';
    const languageCode = 'en';
    const typeOfUse = 'online';

    const defaults = {
      languageCode,
      file,
      typeOfUse,
    };

    function options(overrides = {}) {
      const params = { ...defaults, ...overrides };
      const url = `/attribution/${params.languageCode}/${params.file}/${
        params.typeOfUse
      }/unmodified`;
      return { method: 'GET', url };
    }

    async function subject(overrides = {}) {
      return context.inject(options(overrides));
    }

    beforeEach(() => {
      services.fileData.getFileData.mockResolvedValue(fileInfoMock);
      services.licenses.getLicense.mockResolvedValue(license);
    });

    afterEach(() => {
      services.fileData.getFileData.mockReset();
      services.licenses.getLicense.mockReset();
    });

    it('returns attribution information for the given file', async () => {
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenses.getLicense).toHaveBeenCalledWith(fileInfoMock);

      expect(response.status).toBe(200);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });

    it('returns 500 for any generic error', async () => {
      services.fileData.getFileData.mockImplementation(() => {
        throw new Error('some error');
      });
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenses.getLicense).not.toHaveBeenCalled();
      expect(response.status).toBe(500);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });

    it('returns an error when the file responds with a license we do not know', async () => {
      services.licenses.getLicense.mockImplementation(() => undefined);
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenses.getLicense).toHaveBeenCalledWith(fileInfoMock);
      expect(response.status).toBe(422);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });
  });

  describe('GET /attribution/... (modified)', () => {
    const file = 'File:Foobar.jpg';
    const languageCode = 'en';
    const typeOfUse = 'online';
    const modification = 'cropped';
    const modificationAuthor = 'the great modificator';
    const licenseId = 'cc-by-sa-2.5';

    const defaults = {
      languageCode,
      file,
      typeOfUse,
      modification,
      modificationAuthor,
      licenseId,
    };

    function options(overrides = {}) {
      const params = { ...defaults, ...overrides };
      const url = `/attribution/${params.languageCode}/${params.file}/${
        params.typeOfUse
      }/modified/${params.modification}/${params.modificationAuthor}/${params.licenseId}`;
      return { method: 'GET', url };
    }

    async function subject(overrides = {}) {
      return context.inject(options(overrides));
    }

    beforeEach(() => {
      services.fileData.getFileData.mockResolvedValue(fileInfoMock);
      services.licenseStore.getLicenseById.mockImplementation(() => license);
    });

    afterEach(() => {
      services.fileData.getFileData.mockReset();
      services.licenseStore.getLicenseById.mockReset();
    });

    it('returns attribution information for the given file', async () => {
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenseStore.getLicenseById).toHaveBeenCalledWith(licenseId);

      expect(response.status).toBe(200);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });

    it('returns 500 for any generic error', async () => {
      services.fileData.getFileData.mockImplementation(() => {
        throw new Error('some error');
      });
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenseStore.getLicenseById).not.toHaveBeenCalledWith();
      expect(response.status).toBe(500);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });

    it('returns an error when the requested a license we do not know', async () => {
      services.licenseStore.getLicenseById.mockImplementation(() => undefined);
      const response = await subject();

      expect(services.fileData.getFileData).toHaveBeenCalledWith(file);
      expect(services.licenseStore.getLicenseById).toHaveBeenCalledWith(licenseId);
      expect(response.status).toBe(404);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });

    it('returns an error when the requested a typeOfUse we do not know', async () => {
      const response = await subject({ typeOfUse: 'for the lulz' });

      expect(services.fileData.getFileData).not.toHaveBeenCalledWith();
      expect(services.licenseStore.getLicenseById).not.toHaveBeenCalledWith();
      expect(response.status).toBe(400);
      expect(response.type).toBe('application/json');
      expect(response.payload).toMatchSnapshot();
    });
  });
});
