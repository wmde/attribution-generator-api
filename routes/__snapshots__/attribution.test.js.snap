// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attribution routes GET /attribution/... (modified) returns 500 for any generic error 1`] = `
Object {
  "error": "Internal Server Error",
  "message": "An internal server error occurred",
  "statusCode": 500,
}
`;

exports[`attribution routes GET /attribution/... (modified) returns an error when the requested a license we do not know 1`] = `
Object {
  "error": "Not Found",
  "message": "license-not-found",
  "statusCode": 404,
}
`;

exports[`attribution routes GET /attribution/... (modified) returns an error when the requested a typeOfUse we do not know 1`] = `
Object {
  "error": "Bad Request",
  "message": "Invalid request params input",
  "statusCode": 400,
}
`;

exports[`attribution routes GET /attribution/... (modified) returns attribution information for the given file 1`] = `
Object {
  "attributionHtml": "Photograph by <a href=\\"https://commons.wikimedia.org/wiki/User:Rama\\">Rama</a>, Wikimedia Commons, Cc-by-sa-2.0-fr, <a href=\\"https://upload.wikimedia.org/wikipedia/commons/b/bc/Apple_Lisa2-IMG_1517.jpg\\">Apple_Lisa2-IMG_1517</a>, cropped by the great modificator, <a href=\\"https://creativecommons.org/licenses/by-sa/2.5/legalcode\\" rel=\\"license\\">CC BY-SA 2.5</a>",
  "attributionPlain": "Photograph by Rama, Wikimedia Commons, Cc-by-sa-2.0-fr (https://upload.wikimedia.org/wikipedia/commons/b/bc/Apple_Lisa2-IMG_1517.jpg), „Apple_Lisa2-IMG_1517“, cropped by the great modificator, https://creativecommons.org/licenses/by-sa/2.5/legalcode",
  "licenseId": "cc-by-sa-2.5",
  "licenseUrl": "https://creativecommons.org/licenses/by-sa/2.5/legalcode",
}
`;

exports[`attribution routes GET /attribution/... (unmodified) returns 500 for any generic error 1`] = `
Object {
  "error": "Internal Server Error",
  "message": "An internal server error occurred",
  "statusCode": 500,
}
`;

exports[`attribution routes GET /attribution/... (unmodified) returns an error when the file responds with a license we do not know 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": "validation-error",
  "statusCode": 422,
}
`;

exports[`attribution routes GET /attribution/... (unmodified) returns attribution information for the given file 1`] = `
Object {
  "attributionHtml": "Photograph by <a href=\\"https://commons.wikimedia.org/wiki/User:Rama\\">Rama</a>, Wikimedia Commons, Cc-by-sa-2.0-fr, <a href=\\"https://upload.wikimedia.org/wikipedia/commons/b/bc/Apple_Lisa2-IMG_1517.jpg\\">Apple_Lisa2-IMG_1517</a>, <a href=\\"https://creativecommons.org/licenses/by-sa/2.5/legalcode\\" rel=\\"license\\">CC BY-SA 2.5</a>",
  "attributionPlain": "Photograph by Rama, Wikimedia Commons, Cc-by-sa-2.0-fr (https://upload.wikimedia.org/wikipedia/commons/b/bc/Apple_Lisa2-IMG_1517.jpg), „Apple_Lisa2-IMG_1517“, https://creativecommons.org/licenses/by-sa/2.5/legalcode",
  "licenseId": "cc-by-sa-2.5",
  "licenseUrl": "https://creativecommons.org/licenses/by-sa/2.5/legalcode",
}
`;
